### ISSUE TITLE
application does not run

### ISSUE THREAD

**Issue by mylinttrap** [barrucadu/bookdb#19]
I've run this in my local docker using the python:3.8 image. 

When I run the app, i get this page: The search server is unavailable. Try again in a minute or two.
I did pip install -y requirements-freeze.txt
basically everything that was in the Dockerfile. 

This is after editing serve.py because the redirect breaks the site completely.
from : BASE_URI = os.getenv("BASE_URI", "http://bookdb.nyarlathotep")
to: BASE_URI = os.getenv("BASE_URI", "http://192.168.1.76:8888")

After that I gave up. I dont think this is a repeatable install inside or outside docker. 
You also have a port variabl 9200 that is on localhost but not open or called to be open. so i opened that on my container. 

I like the project idea... hopefully you can make it so that it works correctly. 
I cant spend more time trying to get it to work :(

**Comment by barrucadu** [comment]
You don't need to edit the source to change the `BASE_URI`, you can just set that environment variable (same with `ES_HOST`, `COVER_DIR`, and `ALLOW_WRITES`).

It's looking for an elasticsearch server at `ES_HOST`, which is `http://localhost:9200` by default.  Try this docker-compose file out:

```yaml
version: '3'

services:
  bookdb:
    build: .
    restart: always
    environment:
      ALLOW_WRITES: "1"
      BASE_URI: "http://192.168.1.76:8888"
      COVER_DIR: "/bookdb-covers"
      ES_HOST: "http://db:9200"
    networks:
      - bookdb
    ports:
      - "127.0.0.1:8888:8888"
    volumes:
      - bookdb_covers:/bookdb-covers
    depends_on:
      - db

  db:
    image: elasticsearch:7.6.2
    restart: always
    environment:
      - http.host=0.0.0.0
      - discovery.type=single-node
      - ES_JAVA_OPTS=-Xms1g -Xmx1g
    networks:
      - bookdb
    volumes:
      - bookdb_esdata:/usr/share/elasticsearch/data

networks:
  bookdb:
    external: false

volumes:
  bookdb_covers:
  bookdb_esdata:
```

**Comment by barrucadu** [comment]
You'll also need to run `./src/create-index.py` to set up the Elasticsearch index.

**Comment by mylinttrap** [comment]
Thanks! I'll give it a try soon. Adding this info to the readme.md would help future people like me :D

**Comment by mylinttrap** [comment]
Still not working. 

`version: "3"
networks:
  web:
    external: true
  internal:
    external: false
services:
  bookdb:
    build: ./bookdb/
    container_name: bookdb.domain.com
    hostname: bookdb
    environment:
      - ALLOW_WRITES=1
      - BASE_URI="http://bookdb.domain.com"
      #- BASE_URI: "http://192.168.1.76:8888"
      - COVER_DIR="/bookdb-covers"
      - ES_HOST="http://bookdb-db:9200"
    restart: unless-stopped
    networks:
      - internal
      - web
    labels:
      - traefik.frontend.rule=Host:bookdb.domain.com
      - traefik.docker.network=web
      - traefik.port=8888
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /mnt/bookdb/data/bookdb-covers:/bookdb-covers
    depends_on:
      - bookdb-db

  bookdb-db:
    container_name: bookdb-db
    image: elasticsearch:7.6.2
    restart: unless-stopped
    environment:
      - http.host=0.0.0.0
      - discovery.type=single-node
      - ES_JAVA_OPTS=-Xms1g -Xmx1g
    networks:
      - web
      - internal
    volumes:
      - /mnt/bookdb_esdata:/usr/share/elasticsearch/data
`

I've created the above and it is in a reboot loop. the issue it runs into is that it can't connect to the elasticsearch instance. 
I've confirmed that I can curl and ping from another container. 

`
curl http://bookdb-db:9200
{
  "name" : "4f515263bbb0",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "1HcFRIJBR6mYqrk8KCOsdw",
  "version" : {
    "number" : "7.6.2",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "ef48eb35cf30adf4db14086e8aabd07ef6fb113f",
    "build_date" : "2020-03-26T06:34:37.794943Z",
    "build_snapshot" : false,
    "lucene_version" : "8.4.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
`

these are the env settings in container before it hit boot loop.

HOSTNAME=bookdb
TERM=xterm
ALLOW_WRITES="1"
BASE_URI="http://bookdb.domain.com"
COVER_DIR="/bookdb-covers"
ES_HOST="http://bookdb-db:9200"
LANG=C.UTF-8
GPG_KEY=E3FF2839C0SDFERC084DEBE9B26995E310250568
PYTHON_VERSION=3.8.5
PYTHON_PIP_VERSION=20.2.3
PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/fa7dc83944936bf09a0e4cb5d5ec852c0d256599/get-pip.py
PYTHON_GET_PIP_SHA256=6e0bb0a2c2533361d7f297ed547237caf1b7507f197835974c0dd7eba998c53c
HOME=/home/app


the error i get from container: 


`
Traceback (most recent call last):
  File "/app/serve.py", line 27, in <module>
    es = Elasticsearch([os.getenv("ES_HOST", "http://bookdb-db:9200")])
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/client/__init__.py", line 227, in __init__
    self.transport = transport_class(_normalize_hosts(hosts), **kwargs)
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/transport.py", line 127, in __init__
    self.set_connections(hosts)
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/transport.py", line 182, in set_connections
    connections = list(zip(connections, hosts))
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/transport.py", line 178, in _create_connection
    return self.connection_class(**kwargs)
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 194, in __init__
    self.pool = pool_class(
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 186, in __init__
    ConnectionPool.__init__(self, host, port)
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 80, in __init__
    raise LocationValueError("No host specified.")
urllib3.exceptions.LocationValueError: No host specified.


`

**Comment by mylinttrap** [comment]
To the last comment. I think that because the url is designated as localhost... and is run in container... it has to be an absolute ip. I did an alternate test. where I built up two individual containers. one with elasticsearch and one with your python. 

I was unable to connect over localhost. but as soon as I altered the config to say 192.168.1.71:9200 it was able to connect. 
same with curl. 

so something is funny ... but the app started in this very manual mode. need to figure out the docker-compose part now i suspect. were you able to get this working in a clean environment from scratch? using compose.

**Comment by barrucadu** [comment]
I think you have some extra quoting in

```
    environment:
      - ALLOW_WRITES=1
      - BASE_URI="http://bookdb.domain.com"
      #- BASE_URI: "http://192.168.1.76:8888"
      - COVER_DIR="/bookdb-covers"
      - ES_HOST="http://bookdb-db:9200"
```

If I change bookdb to just `print([os.getenv("ES_HOST", "http://localhost:9200")])`, then that gives me:

```
bookdb.domain.com | ['"http://bookdb-db:9200"']
bookdb.domain.com exited with code 0
```

See the extra `"`s inside the `'`s

**Comment by mylinttrap** [comment]
I see. I'll give that a shot. 

How do I add books and authors?
This page shows up when i replace search for edit after the baseurl
http://192.168.3.85:8888/edit


The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.

Click here to continue

**Comment by barrucadu** [comment]
It's `/add`, there should be a link in the footer.

**Comment by mylinttrap** [comment]
Not seeing it. add.html does not exist in the templates dir

**Comment by barrucadu** [comment]
It uses the edit template but the route is `/add` (edit routes are like `/book/9780486272634/edit`).  It shows up in the footer if `ALLOW_WRITES=1` is in the environment.

**Comment by mylinttrap** [comment]
nice. did an exprot ALLOW_WRITES=1 on the manual dockers I have running. 
Sill playing with the compose install . I have some variables to sort out there. 

The compose has not been able to get up and running with the url. And the individual dockers are running off IP configurations. but the app is up and running in the IP Scenario and I was able to get add url working. but ran into another error


`
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "./serve.py", line 505, in add
    return do_create_book(request)
  File "./serve.py", line 396, in do_create_book
    es.create(index="bookdb", id=bId, body=candidate)
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/client/utils.py", line 92, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/client/__init__.py", line 340, in create
    return self.transport.perform_request(
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/transport.py", line 355, in perform_request
    status, headers_response, data = connection.perform_request(
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 248, in perform_request
    self._raise_error(response.status, raw_data)
  File "/usr/local/lib/python3.8/site-packages/elasticsearch/connection/base.py", line 243, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
elasticsearch.exceptions.AuthorizationException: AuthorizationException(403, 'cluster_block_exception', 'index [bookdb] blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];')
192.168.3.13 - - [16/Sep/2020 16:37:18] "POST /add HTTP/1.1" 500 -
`

**Comment by barrucadu** [comment]
I've not seen that error before, and I'm successfully using Elasticsearch 7.6.2 locally.  Are there any errors from the elasticsearch container?

**Comment by mylinttrap** [comment]
might have found it... got a warning.

{"type": "server", "timestamp": "2020-09-16T16:43:40,643Z", "level": "WARN", "component": "o.e.c.r.a.DiskThresholdMonitor", "cluster.name": "docker-cluster", "node.name": "56c4e4278c7e", "message": "flood stage disk watermark [95%] exceeded on [ZlBAld0KSvS52pldfbh3LA][56c4e4278c7e][/usr/share/elasticsearch/data/nodes/0] free: 826.5mb[2.8%], all indices on this node will be marked read-only", "cluster.uuid": "bm0I6PZ7REqwqip7eV6-DQ", "node.id": "ZlBAld0KSvS52pldfbh3LA"  }

**Comment by barrucadu** [comment]
Right, there may be a way to configure elasticsearch to work when over 85% of the disk is full, or it might just be easier to free up some space

**Comment by mylinttrap** [comment]
https://selleo.com/til/posts/esrgfyxjee-how-to-fix-elasticsearch-forbidden12index-read-only

The fix! ^^

**Comment by mylinttrap** [comment]
```
curl -XPUT -H "Content-Type: application/json" http://localhost:9200/_cluster/settings -d '{ "transient": { "cluster.routing.allocation.disk.threshold_enabled": false } }'
curl -XPUT -H "Content-Type: application/json" http://localhost:9200/_all/_settings -d '{"index.blocks.read_only_allow_delete": null}'

```

**Comment by mylinttrap** [comment]
Thanks! I got it up and working in the manual config mode. Going to take a break and then come back to the docker-compose mode. Thanks! 

Will be trying to use the CLI to input records to the db. http POST should be sufficient with curl/wget I think... unless you have a better idea. 

Thanks for the help on on getting it running!

**Comment by mylinttrap** [comment]
Of course... it might be easy to take the serve.py script and scrounge out the code to make it a cli tool. bood-db-add.py or something. :D

**Comment by barrucadu** [comment]
`curl` is probably the easiest way for now, though you could also import a json file with the `create-index.py` script, it can take an input of the form

```json
{
   "0000001270921" : {
      "bucket" : "haddon",
      "cover_image_mimetype" : "image/jpeg",
      "created_at" : "2020-05-07T21:02:08",
      "display_title" : "The Curious Incident of the Dog in the Night-Time",
      "has_been_read" : false,
      "holdings" : [
         {
            "location_uuid" : "f256ed66-4c09-4207-86de-adc8e9fb86ec"
         }
      ],
      "people" : {
         "authors" : [
            "Mark Haddon"
         ]
      },
      "title" : "The Curious Incident of the Dog in the Night-Time",
      "updated_at" : "2020-05-07T21:02:08"
   },
   "0006716636" : {
      "bucket" : "lewis",
      "category_uuid" : "ea4e9b04-3ebb-4ff3-b387-d8bfc603f58a",
      "cover_image_mimetype" : "image/jpeg",
      "created_at" : "2020-05-07T21:02:08",
      "display_title" : "The Lion, the Witch and the Wardrobe",
      "has_been_read" : true,
      "holdings" : [
         {
            "location_uuid" : "be60be7b-a10f-42e1-8769-d43f12cad02d"
         }
      ],
      "people" : {
         "authors" : [
            "C. S. Lewis"
         ]
      },
      "title" : "The Lion, the Witch and the Wardrobe",
      "updated_at" : "2020-05-07T21:02:08"
   },
...
}
```

That wouldn't handle uploading cover images... but you could copy those to the right location in the container I suppose.

I use a combination of `dump-index.py` / `create-index.py`, `docker cp`, and `scp` to sync bookdb data from my local instance (which is writable) to https://bookdb.barrucadu.co.uk/ (which is read-only).

**Comment by barrucadu** [comment]
I expect you'll also want to edit `config/uuids.yaml` file so you've got the locations and categories you want.

**Comment by mylinttrap** [comment]
Ha. I got the docker-compose to work with my traefik config. Turns out it was not loading in the browser because my debug machine didn't have traefik installed on it this last rebuild. 

Good thing though. because with traefik and flask it strips the css for the site. I've run into this before. usually when there is a backend proxy/loadbalancer built into the app. i think flask might have something like that going on and the headers are getting mangled. unsure, though it is interesting . 

for now i'll be using it on url:8888 since it works and looks as it should.
